### PR TITLE
Smart library demo

### PR DESCRIPTION
**Pull request by FenixFly**
Add Python demo "Smart Library".

It involves the registration of the reader; authorization of the reader through face recognition; 
receiving and returning books by recognizing QR codes generated for each book in the library.

### CODE REVIEW

**Comment by eizamaliev** [comment]
Please, retarget to `opencv:develop`

**Comment by eizamaliev** [comment]
Please, add demo to `demos/README.md`

**Comment by eizamaliev** [comment]
Also take a look at [contribution guide](https://github.com/opencv/open_model_zoo/blob/develop/CONTRIBUTING.md#demo), section **DEMO**

File: demos/python_demos/smart_library_demo/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+extra statictics in console, like list of registered readers, full list of books in the library,
+history of borrowing books. To change information in console press `f` on keyboard. To exit press `q`.
+
+### Installation and dependencies
+
+The demo depends on:
+- OpenVINO toolkit (2019R3 or newer)
+- Python (any of 2.7+ or 3.4+, which is supported by OpenVINO)
+- OpenCV (>=3.4.0)
+
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/README.md]: We don't need to specify this. OMZ is supposed to work with OpenVINO and OpenCV of versions it is shipped with.

File: demos/python_demos/smart_library_demo/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+Models:
+  -fr FREC           Required. Type of recognizer. Available DNN face
+                     recognizer - DNNfr
+  -m_rd RDMODEL      Required. Path to .xml file
+  -fd FDDET          Required. Type of detector. Available DNN face detector -
+                     DNNfd
+  -m_fd FDMODEL      Required. Path to .xml file
+  -lm LMDET          Required. Type of detector. Available DNN landmarks
+                     regression - DNNlm
+  -m_lm LMMODEL      Required. Path to .xml file
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/README.md]: The demo prints a different message.

File: demos/python_demos/smart_library_demo/createQRCodes.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                        book['author'] + ' ' + book['publisher'] + ' ' +
+                        str(book['year']))
+                qr = gen.makeQR(strData)
+                print(strData)
+                qr.save(args.out + '/' + str(book['id']) + '.png')
+    else:
+        print('File or directory not exists!')
+
+if __name__ == '__main__':
+    sys.exit(main() or 0)
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/createQRCodes.py]:
```suggestion
    sys.exit(main() or 0)
    
```

File: demos/python_demos/smart_library_demo/library.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        },
+        {
+            "id": 2,
+            "title": "book 2",
+            "author": "Author 1, Author 2",
+            "publisher": "pub",
+            "year": 2017
+        } 
+    ]
+}
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/library.json]:
```suggestion
}

```

File: demos/python_demos/smart_library_demo/models.lst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# This file can be used with the --list option of the model downloader.
+face-detection-retail-????
+face-reidentification-retail-????
+landmarks-regression-retail-????
+
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/models.lst]:
```suggestion

```

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+import argparse
+import numpy as np
+import cv2 as cv
+import json
+
+sys.path.insert(0,'src/modules')
+import face_recognizer as fr 
+import book_recognizer as br
+
+sys.path.insert(0, 'src/infrastructure')
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]: This solutions works only if the working directory is `open_model_zoo/demos/python_demos/smart_library_demo`. It is possible to import modules like `from src.modules import face_recognizer as fr`.

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+from DynamicDatabase import *
+
+uknownID = -1
+count = 0
+
+def CreateArgparse():
+    parser = argparse.ArgumentParser(description='Smart Library Sample') 
+    parser.add_argument('-fr', type = str,
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]: Such keys expect only one specific option, so there is no reason to introduce them

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+                        dest = 'fRec', required=True,
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]:
```suggestion
                        required=True,
```
The value of dest is normally inferred from the option strings. The same goes for the rest keys.

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+from DynamicDatabase import *
+
+uknownID = -1
+count = 0
+
+def CreateArgparse():
+    parser = argparse.ArgumentParser(description='Smart Library Sample') 
+    parser.add_argument('-fr', type = str,
+                        dest = 'fRec', required=True, 
+                        help = 'Required. \
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]:
```suggestion
                        help='Required. \
```
And for other keyword arguments.

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                        Type of recognizer. \
+                        Available DNN face recognizer - DNNfr')
+    parser.add_argument('-m_rd', type = str, 
+                        dest = 'rdModel', required=True,  
+                        help = 'Required. Path to .xml file')
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]:
```suggestion
                        help = 'Required. Path to .xml file with a face reidentification model')
```
You should clarify what model is expected. The same goes for other keys.

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    parser.add_argument('-h_lm', type = int, default = 48,
+                        dest = 'lmHeight',
+                        help = 'Optional. Image height to resize') 
+    parser.add_argument('-br', type = str,  default = 'QR',
+                        dest = 'br',
+                        help = 'Optional. Type - QR')
+    parser.add_argument('-lib', type = str, default = 'library.json',
+                        dest = 'lib',
+                        help = 'Optional. Path to library.') 
+    parser.add_argument('-w', type = int, default = 0,
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]: The name of the key is usually `'-i', '--input'`, it usually accepts a string, and tries casting it to int before passing it to VideoCapture, to support other sources.

File: demos/python_demos/smart_library_demo/smart_library_demo.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+def CreateLibrary(libPath, DB):
+    with open(libPath, 'r', encoding='utf-8') as lib:
+        data = json.load(lib)
+    for book in data['books']:
+        DB.addBook(book['id'], book['title'],  book['author'],
+                   book['publisher'], book['year'])
+
+def PutText(img, text, pos, ix, iy, font, color, scale, thickness, rect = 1):
+
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/smart_library_demo.py]:
```suggestion

```

File: demos/python_demos/smart_library_demo/src/infrastructure/DynamicDatabase.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        print('{:<10}{:<10}{:<20}{:<40}{:<20}{:<20}'.format('User ID', 'Book ID', 'First name',
+             'Title',  'Borrow date','Return date'))
+        for bbook in self.BBooks:
+            for book in self.Books:
+                for user in self.Users:
+                    if bbook.userID == user.user_id and bbook.bookID == book.book_id:
+                        print('{:<10}{:<10}{:<20}{:<40}{:<20}{:<20}'.format(user.user_id, book.book_id,
+                                   user.first_name, book.title, bbook.bdate, bbook.rdate))
+                        
+
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/src/infrastructure/DynamicDatabase.py]: Please, remove extra line and trailing white spaces.

File: demos/python_demos/smart_library_demo/src/infrastructure/Entities/Author.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+        self.author_id = author_id
+        self.first_name = first_name
+        self.last_name = last_name
+        self.middle_name = middle_name
+        
+    def _print(self):
+        print(self.author_id)
+        print(self.first_name)
+        print(self.last_name)
+        print(self.middle_name)
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/src/infrastructure/Entities/Author.py]:
```suggestion
        print(self.middle_name)

```

File: demos/python_demos/smart_library_demo/src/infrastructure/Entities/Book.py [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+        print(self.book_id)
+        print(self.file_path)
+        print(self.title)
+        print(self.year)
+        print(self.publisher)
+        print("AUTHORS:")
+        i = 0
+        for elem in self.authors:
+            (self.authors[i])._print()
+            i = i + 1
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/src/infrastructure/Entities/Book.py]:
```suggestion
            i = i + 1

```

File: demos/python_demos/smart_library_demo/src/infrastructure/Entities/User.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+        self.first_name = first_name
+        self.last_name = last_name
+        self.middle_name = middle_name
+        
+    def _print(self):
+        print(self.user_id)
+        print(self.phone)
+        print(self.first_name)
+        print(self.last_name)
+        print(self.middle_name)
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/src/infrastructure/Entities/User.py]:
```suggestion
        print(self.middle_name)

```

File: demos/python_demos/smart_library_demo/src/modules/book_recognizer.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+        self.objects = pyzbar.decode(gray)
+
+        # Analise results
+        ans = ""
+        for obj in self.objects:
+            if obj.type == 'QRCODE':
+                ans = obj.data.decode('utf-8')
+
+        # Return decode information
+        return ans
```
Comment (Wovchena) [inline demos/python_demos/smart_library_demo/src/modules/book_recognizer.py]:
```suggestion
        return ans

```

**Pull request closed without merging**